File: app/src/main/java/com/kickstarter/ui/activities/ProjectNotificationSettingsActivity.java
package com.kickstarter.ui.activities;

import android.os.Bundle;
import android.util.Pair;

import com.kickstarter.R;
import com.kickstarter.libs.BaseActivity;
import com.kickstarter.libs.qualifiers.RequiresActivityViewModel;
import com.kickstarter.ui.adapters.ProjectNotificationSettingsAdapter;
import com.kickstarter.viewmodels.ProjectNotificationSettingsViewModel;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.Bind;
import butterknife.BindString;
import butterknife.ButterKnife;
import rx.android.schedulers.AndroidSchedulers;

import static com.kickstarter.extensions.ActivityExtKt.showSnackbar;
import static com.kickstarter.libs.utils.TransitionUtils.slideInFromLeft;

@RequiresActivityViewModel(ProjectNotificationSettingsViewModel.ViewModel.class)
public final class ProjectNotificationSettingsActivity extends BaseActivity<ProjectNotificationSettingsViewModel.ViewModel> {
  protected @Bind(R.id.project_notification_settings_recycler_view) RecyclerView recyclerView;

  protected @BindString(R.string.general_error_something_wrong) String generalErrorString;

  @Override
  protected void onCreate(final @Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.project_notification_settings_layout);
    ButterKnife.bind(this);

    final ProjectNotificationSettingsAdapter adapter = new ProjectNotificationSettingsAdapter();
    this.recyclerView.setAdapter(adapter);
    this.recyclerView.setLayoutManager(new LinearLayoutManager(this));

    this.viewModel.outputs.projectNotifications()
      .compose(bindToLifecycle())
      .observeOn(AndroidSchedulers.mainThread())
      .subscribe(adapter::projectNotifications);

    this.viewModel.outputs.unableToFetchProjectNotificationsError()
      .compose(bindToLifecycle())
      .observeOn(AndroidSchedulers.mainThread())
      .subscribe(__ -> showSnackbar(this.recyclerView, this.generalErrorString));
  }

  @Override
  protected void onDestroy() {
    super.onDestroy();
    this.recyclerView.setAdapter(null);
  }

  @Override
  protected @Nullable Pair<Integer, Integer> exitTransition() {
    return slideInFromLeft();
  }
}

File: app/src/main/java/com/kickstarter/ui/activities/ProjectNotificationSettingsActivity.kt
package com.kickstarter.ui.activities

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.viewModels
import androidx.recyclerview.widget.LinearLayoutManager
import com.kickstarter.R
import com.kickstarter.databinding.ProjectNotificationSettingsLayoutBinding
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.libs.utils.extensions.getEnvironment
import com.kickstarter.ui.adapters.ProjectNotificationSettingsAdapter
import com.kickstarter.ui.extensions.showSnackbar
import com.kickstarter.utils.WindowInsetsUtil
import com.kickstarter.viewmodels.ProjectNotificationSettingsViewModel
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable

class ProjectNotificationSettingsActivity : ComponentActivity() {
    private lateinit var binding: ProjectNotificationSettingsLayoutBinding
    private lateinit var viewModelFactory: ProjectNotificationSettingsViewModel.Factory
    private val viewModel: ProjectNotificationSettingsViewModel.ProjectNotificationSettingsViewModel by viewModels {
        viewModelFactory
    }

    private val disposables = CompositeDisposable()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        getEnvironment()?.let { env ->
            viewModelFactory = ProjectNotificationSettingsViewModel.Factory(env)
        }
        binding = ProjectNotificationSettingsLayoutBinding.inflate(layoutInflater)
        WindowInsetsUtil.manageEdgeToEdge(
            window,
            binding.root,

        )
        setContentView(binding.root)

        val adapter = ProjectNotificationSettingsAdapter()

        binding.projectNotificationSettingsRecyclerView.adapter = adapter
        binding.projectNotificationSettingsRecyclerView.layoutManager = LinearLayoutManager(this)

        viewModel.outputs.projectNotifications()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { adapter.projectNotifications(it) }
            .addToDisposable(disposables)

        viewModel.outputs.unableToFetchProjectNotificationsError()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe {
                showSnackbar(
                    binding.projectNotificationSettingsRecyclerView,
                    getString(R.string.general_error_something_wrong)
                )
            }
            .addToDisposable(disposables)
    }

    override fun onDestroy() {
        super.onDestroy()
        disposables.clear()
        binding.projectNotificationSettingsRecyclerView.adapter = null
    }
}
